File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/neo/promisehandlers/QueryStatisticsHandler.scala
package edu.uci.ics.amber.engine.architecture.worker.neo.promisehandlers

import edu.uci.ics.amber.engine.architecture.worker.WorkerStatistics
import edu.uci.ics.amber.engine.architecture.worker.neo.WorkerAsyncRPCHandlerInitializer
import edu.uci.ics.amber.engine.architecture.worker.neo.promisehandlers.QueryStatisticsHandler.QueryStatistics
import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.{CommandCompleted, ControlCommand}

object QueryStatisticsHandler{
  final case class QueryStatistics() extends ControlCommand[WorkerStatistics]
}


trait QueryStatisticsHandler {
  this:WorkerAsyncRPCHandlerInitializer =>

  registerHandler{
    (msg: QueryStatistics, sender) =>
      val (in, out) = dataProcessor.collectStatistics()
      val state = stateManager.getCurrentState
      WorkerStatistics(state, in, out)
  }

}

File: core/amber/src/main/python/core/architecture/handlers/control/query_statistics_handler.py
from proto.edu.uci.ics.amber.engine.architecture.worker import (
    QueryStatisticsV2,
    WorkerStatistics,
)
from core.architecture.handlers.control.control_handler_base import ControlHandler
from core.architecture.managers.context import Context


class QueryStatisticsHandler(ControlHandler):
    cmd = QueryStatisticsV2

    def __call__(self, context: Context, command: QueryStatisticsV2, *args, **kwargs):
        stats = context.statistics_manager.get_statistics()
        return WorkerStatistics(
            worker_state=context.state_manager.get_current_state(),
            input_tuple_count=stats[0],
            output_tuple_count=stats[1],
            data_processing_time=stats[2],
            control_processing_time=stats[3],
            idle_time=stats[4],
        )
